src/fetch/contracts.py

```diff
@@ -66,21 +60,34 @@ def _limited_is_contract(self, addresses: list[str]) -> dict[str, bool]:
         results = {}
         for result_dict in response.json():
             try:
-                results[addresses[result_dict['id']]] = result_dict['result'] != "0x"
+                results[addresses[result_dict['id']]] = result_dict['result']
             except KeyError as err:
                 raise IOError(
-                    f"Request for address \"{addresses[result_dict['id']]}\" "
+                    f"Request for code at address \"{addresses[result_dict['id']]}\" "
                     f"failed with response {result_dict}"
                 ) from err
         return results
 
     def load_from_file(self, file: File) -> dict[str, bool]:
         """Loads results dict from a file containing known contract addresses"""
-        print(f"Loading Contracts from {file.name}")
+        print(f"loading contracts from {file.name}")
         with open(file.filename(), 'r', encoding='utf-8') as txt_file:
             contracts = set(txt_file.read().splitlines())
         return {addr: addr in contracts for addr in self.addresses}
 
+    def batch_call(
+            self,
+            addresses: list[str],
+            func: Callable[[list[str]], dict[str, int | str]]
+    ) -> dict[str, int | str]:
```

Comment:
To make `batch_call` return the right type and be as general purpose as possible we need to use generics here.

This way the return type of `batch_call` doesn't have to be the union of the return types of all functions we want to call it with (I think this is what you wanted to achieve with `dict[str, int | str]`).
Instead `batch_call` would return `dict[str, str]` if called with `_get_code_at` and return `dict[str, int]` if called with `_limited_balances`.

This can be done with [TypeVar](https://docs.python.org/3/library/typing.html#typing.TypeVar`) and the signature of `batch_call` would look like this:
```python
    def batch_call(
            self,
            addresses: list[A],
            func: Callable[[list[A]], dict[K, V]]
    ) -> dict[K, V]:
```

Reply:
Thanks! This is great. Will do.